### DISCUSSION TITLE
Port config/init confusion, whats the recomended way?

### DISCUSSION THREAD

**Discussion by duckylotl** [hathach/tinyusb#2558] [Q&A] [answered]
Looking at the examples they seem to all use `tud_init` and/or `tuh_init` instead of `tusb_init.` Why call each init individually if there is a function that takes care of it depending on the config?

The commit that made that change to seemingly allow configurable port numbers (with `BOARD_TUD_THPORT`).

That should have been possible with the `TUD_OPT_RHPORT` option as well should it not?
Which does not seem to be set by any config, but rather by proxy through this combined config `CFG_TUSB_RHPORT0_MODE`.

The newer style of config.h does no longer use the combined `CFG_TUSB_RHPORT0_MODE` configurations (which would setup `TUD_OPT_RHPORT`, `CFG_TUD_ENABLED`/`CFG_TUH_ENABLED`, `CFG_TUD_MAX_SPEED`) instead it sets each of these individually.

Except `TUD_OPT_RHPORT`, here it only sets `BOARD_TUD_RHPORT.`

Would it not make sense to just also set `TUD_OPT_RHPORT` directly instead of adding another config variable?
Or implicitly setting this when not defined in tusb_option?
```c
#ifndef TUD_OPT_RHPORT
#define TUD_OPT_RHPORT      BOARD_TUD_RHPORT
#endif
```
Then could also again use the generic `tusb_init` again.

OR is `tusb_init`, `TUD_OPT_RHPORT`, and `CFG_TUSB_RHPORT0_MODE` all deprecated and should no longer be used?
The use of `tusb_init` does seem to limit oneself to the use of only a single port, while the separate init of host and device stack seem to allow using different ports for each. (So the above fallback definition probably has to be more complicated to account for host and device port defines)

So is the recommended way no longer using the combined config `CFG_TUSB_RHPORT0_MODE,` and initializing the host/device stack explicitly?

Background: I'm trying to fix/update tinyusb examples in Chibios_Contrib, these are really old and use the old style of config.

EDIT: Just saw that the getting_started reference doc (still) advises to use an example config and use of `tusb_init`, but current example `tusb_config.h` will result `tusb_init` in doing nothing because of the no longer defined `TUD_OPT_RHPORT`.

**Comment by hathach** [comment, marked as answer]
the old CFG_TUSB_RHPORT0_MODE style requires the port must be hard coded to host/device. tud_init()/tuh_init() allow user to dynamically init an roothub port as device/host as needed, and deinit() it if want to (save/power/switch role, configuration etc..). deinit() is not implemented for device, but is added for host #2530. 

In summary, it make it easier to switch role dynamically on multiple roothub port controller. If you don't need it, you don't have to use it. tusb_init() can still be invoked to initialize both host + device

**Reply by duckylotl** [reply]
I think I have a better idea now. The new style of initialization definitely gives more flexibility. Even if I might not use its features I will go with the `tud_init`/`tuh_init` from now on since that is more in line with the current state of examples.

I would like however to point out that the getting_started doc still references the `tusb_init` function in combination with example configurations. But none of the examples are compatible with the use of `tusb_init`. Maybe it would make sense to update the doc and promote the `tud_init`/`tuh_init` instead?
In fact the only things that still use `tusb_init` (and the old style config) are two unit tests (test_msc_device and test_usbd).

**Reply by hathach** [reply]
Yeah, we should. But I am always busy fixing other bugs. If you could please make an pr
